### PR TITLE
fix: Delete artifact bundle's assemble status when it is deleted

### LINKED ISSUE #1724: Unable to properly delete uploaded artifacts ("Nothing to upload, all files are on the server")
### Issue

If I delete all uploaded artifact bundles on sentry.io and then try to upload the same artifacts (i.e. without making any changes that would produce a new bundle/"Bundle ID") I get "Nothing to upload, all files are on the server". However, my uploaded events are not being retraced due to "Source Maps not uploaded ", so the deleted artifacts are only partially there, i.e. they are blocking me from uploading new ones for the same bundle but cannot be inspected or deleted or used for retracing. That is pretty confusing. 

### Environment

sentry-cli version 1.75.2
MacOS

### Steps to Reproduce

1. `npx react-native@latest init SentryTest`
2. `sentry-wizard -i reactNative -p android`
3. Answer the four prompts (organization slug, project slug, DSN, auth token)
4. `npx react-native run-android --mode release`
5. Delete all uploaded artifact bundles from sentry.io https://test-32ecaf969.sentry.io/settings/projects/sentry-test/source-maps/artifact-bundles/
Screenshot 2023-08-29 at 14 21 44
6. Rerun `npx react-native run-android --mode release`

### Actual Result

Nothing is uploaded even though there are no artifact (or release) bundles present in sentry.io:

```
> Task :app:createBundleReleaseJsAndAssets_SentryUpload_com.sentrytest@1.0+1_1
Processing react-native sourcemaps for Sentry upload.
> Analyzing 2 sources
> Rewriting sources
> Adding source map references
Uploading sourcemaps for release com.sentrytest@1.0+1 distribution 1
> Bundled 2 files for upload
> Bundle ID: dea28e39-4391-56d5-8879-f0b9dc30a906
> Nothing to upload, all files are on the server
```

### Expected result

I would expect that deleting the artifact bundles from sentry.io would allow me to upload them again. However, it looks like the artifacts are somewhat still present somewhere where I cannot properly delete them from.

### PR DESCRIPTION
**Pull request by loewenheim**
Assemble statuses persist in the cache for a while, which can result in the artifact bundle being reported to `sentry-cli` as already existing after deletion.

Should fix https://github.com/getsentry/sentry-cli/issues/1724.

### CODE REVIEW

**Review by Swatinem — COMMENTED** [review]
this also needs a test that ideally reproduces the problem before fixing it.

File: src/sentry/api/endpoints/artifact_bundles.py [hunk 1]
```diff
@@ -177,6 +178,12 @@ def delete(self, request: Request, project) -> Response:
                             # element, so we just take it out and delete it. This deletion will cascade delete every
                             # connected entity.
                             artifact_bundle.delete()
+                            delete_assemble_status(
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply src/sentry/api/endpoints/artifact_bundles.py]: Alright, I moved the status deletion to the artifact bundle `post_delete` hook.

**Comment by loewenheim** [comment]
> this also needs a test that ideally reproduces the problem before fixing it.

Yup, I'm thinking about how to do this 🤔

**Comment by loewenheim** [comment]
I rebased this so that the first commit just adds a failing test case (actually, the test case itself already existed, I just added a check for the assemble status). The deletion of the assemble status is now slightly more defensive. We only call `delete_assemble_status` if both the organization id and the checksum are present. I'm not sure how they couldn't be, but it does happen in one of the other tests.

**Review by Swatinem — APPROVED** [review]

**Pull request merged**